### PR TITLE
Functional tests

### PR DESCRIPTION
**Pull request by rabbl**
I have added some functional tests.
Every api-command has his own test.
Every test starts with a clean and empty eventstore.
Some api-commands needs that a user or a todo already exists, so this necessary requests are made before.

@codeliner gave the idea to attach listener and get the events from there.
I had some problems with this, so at the moment I'm loading the events from the stream by name.

Any proposals for improvement are welcome, for me this is the first contribution to a prooph-repo.

### CODE REVIEW

File: test/AppBundle/Controller/DefaultControllerTest.php [hunk 1]
```diff
@@ -9,10 +9,8 @@ class DefaultControllerTest extends WebTestCase
     public function testIndex()
     {
         $client = static::createClient();
-
-        $crawler = $client->request('GET', '/');
-
+        $client->request('GET', '/');
         $this->assertEquals(200, $client->getResponse()->getStatusCode());
-        $this->assertContains('Welcome to Symfony', $crawler->filter('#container h1')->text());
+        #$this->assertContains('Welcome to Symfony', $crawler->filter('#container h1')->text());
```
Comment (oqq) [inline test/AppBundle/Controller/DefaultControllerTest.php]: remove this line

File: test/AppBundle/Controller/ApiCommandAddDeadlineToTodoTest.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
+<?php
```
Comment (oqq) [inline test/AppBundle/Controller/ApiCommandAddDeadlineToTodoTest.php:1]: missing `declare(strict_types=1)` directive

File: test/AppBundle/Controller/ApiCommandAddReminderToTodoTest.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
+<?php
```
Comment (oqq) [inline test/AppBundle/Controller/ApiCommandAddReminderToTodoTest.php:1]: missing declare(strict_types=1) directive

File: test/AppBundle/Controller/ApiCommandMarkTodoAsDoneTest.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
+<?php
```
Comment (oqq) [inline test/AppBundle/Controller/ApiCommandMarkTodoAsDoneTest.php:1]: missing declare(strict_types=1) directive

File: test/AppBundle/Controller/ApiCommandPostTodoTest.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?php
```
Comment (oqq) [inline test/AppBundle/Controller/ApiCommandPostTodoTest.php:1]: missing declare(strict_types=1) directive

File: test/AppBundle/Controller/ApiCommandRegisterUserTest.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
+<?php
```
Comment (oqq) [inline test/AppBundle/Controller/ApiCommandRegisterUserTest.php:1]: missing declare(strict_types=1) directive

File: test/AppBundle/Controller/ApiCommandReopenTodoTest.php [hunk 7]
```diff
@@ -0,0 +1,58 @@
+<?php
```
Comment (oqq) [inline test/AppBundle/Controller/ApiCommandReopenTodoTest.php:1]: missing declare(strict_types=1) directive

File: test/AppBundle/Controller/ControllerBaseTestCase.php [hunk 8]
```diff
@@ -0,0 +1,178 @@
+<?php
```
Comment (oqq) [inline test/AppBundle/Controller/ControllerBaseTestCase.php:1]: missing declare(strict_types=1) directive

File: test/AppBundle/Controller/ControllerBaseTestCase.php [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+            array(),
+            array(),
+            json_encode($payload)
+        );
+    }
+
+    /**
+     * @param Uuid $assigneeId
+     * @param Uuid $todoId
+     * @param string $todoDescription
```
Comment (oqq) [inline test/AppBundle/Controller/ControllerBaseTestCase.php]: unnecessary doc annotations

File: test/AppBundle/Controller/ControllerBaseTestCase.php [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        $input->setInteractive(false);
+        $command->run($input, new ConsoleOutput(ConsoleOutput::VERBOSITY_QUIET));
+
+        self::$client = static::createClient();
+    }
+
+    /**
+     * @param Uuid $id
+     * @param string $name
+     * @param string $email
```
Comment (oqq) [inline test/AppBundle/Controller/ControllerBaseTestCase.php]: unnecessary doc annotations

**Review by oqq — COMMENTED** [review]
Some comments from me.

You could remove the first empty line between class declaration and first method on every test class.

Maybe all classes can be final.

**Comment by rabbl** [comment]
I have made the changes, following the hints of @oqq.

**Comment by codeliner** [comment]
@rabbl  :+1: 

One question: What does the WebTestCase do or better the `client`? Are real requests performed?
Attaching listeners to the event store only works if the instance referenced by `$this->store` is the same instance as the one used in `self::client->request(...)`
Anyway, loading the stream and check recored events is a good idea for this kind of tests.

**Comment by rabbl** [comment]
[The client simulates an HTTP-Client and makes Requests to the Application.](https://symfony.com/doc/current/testing.html#working-with-the-test-client)

The `$client = static::createClient();` starts the application and implements the containerAwareInterface. @codeliner, your example code can be implemented and works, but I had the simple problem that I could not extract the recorded events in an easy way. For sure you have an answer for that. I can prepare this in a separate branch that you can have a look at it. 

Your way is easier, because in this case there is no need to drop/create/migrate before each test.

**Comment by codeliner** [comment]
ah ok thx, in this case you can use the same technique like we use in the event store tests: https://github.com/prooph/event-store/blob/master/tests/EventStoreTest.php#L41

And it is maybe a good idea to replace the DoctrineAdapter with the [InMemoryAdapter](https://github.com/prooph/event-store/blob/master/src/Adapter/InMemoryAdapter.php) to improve performance of the tests

**Comment by rabbl** [comment]
With the example from the EventStoreTest it works, great.
With the other example `$recordedEvents`was from type `\ArrayIterator` and not from type array.

Which way do you prefer? Retrieve the events from the stream or record it via a listener?

In my opinion the functional tests should focus on the api-call with correct and wrong (to be implemented) payloads and check if the expected result has happened. 
Is the expected result the emitted event or the stored event in the database?

**Comment by codeliner** [comment]
> is the expected result the emitted event or the stored event in the database?

It really depends on the purpose of the test here. In an End-to-End test I'm more interested in the emitted event because this would be the trigger for a read model update or a process manager.
But it is also very important that events are written in stone before they are communicated. 
The latter is guaranteed by our unit tests so I'd say we are more interested in the emitted events here.

You could also go a step further and attach a listener to the event bus because that is the output port of the application. The `commit.post` action event of the event store is part of the plugin api but normally you don't interact with it directly but use the [EventPublisher](https://github.com/prooph/event-store-bus-bridge/blob/master/docs/event_publisher.md#event-publishing) in combination with the event-bus to send all recorded events to your read model and process managers.

The event publisher is configured [here](https://github.com/prooph/proophessor-do-symfony/blob/master/src/AppBundle/Resources/config/event_store.yml#L29)

and the configuration also tells you that the event publisher forwards all events to `@prooph_service_bus.todo_event_bus`

Finally, you can use a similar listener technique to check that the events are published. See this example: https://github.com/prooph/service-bus/blob/master/tests/EventBusTest.php#L47

**Comment by rabbl** [comment]
At the moment the [configuration](https://github.com/rabbl/proophessor-do-symfony/blob/functionalTestsWithListener/test/AppBundle/Controller/ControllerBaseTestCase.php#L37) does not give the expected results.
Look at [this test](https://github.com/rabbl/proophessor-do-symfony/blob/functionalTestsWithListener/test/AppBundle/Controller/ApiCommandPostTodoTest.php#L31).

I'm expecting two events, or at least the event from the last event (todoWasPosted). 
But there is only the UserWasRegistered event in the array of recorded event. 

Ok, got it. 
I think I have to instantiate a new client with every request.

**Comment by codeliner** [comment]
@rabbl can you do me a favor and try this:

```php
    public function test_command_post_todo_adds_TodoWasPosted_event_to_recordedEvents()
    {
        $this->assertSame($this->eventStore, $this->client->getContainer()
            ->get('prooph_event_store.todo_store'));
        $this->assertCount(2, $this->recordedEvents);
        $this->assertInstanceOf(UserWasRegistered::class,$this->recordedEvents[0]);
        $this->assertInstanceOf(TodoWasPosted::class,$this->recordedEvents[1]);
    }
```

**Comment by codeliner** [comment]
I bet that symfony resets all instances in the container after each request.

At least I don't see another issue with your test other than that. If the assertion does not fail only debugging would help

**Comment by rabbl** [comment]
I have updated the tests, adding listener and testing the recorded events.
At the beginning of all tests the database will be dropped/created/migrated silently.

**Pull request merged**